crates/wasmtime/src/func.rs

```diff
@@ -1254,35 +1254,33 @@ pub(crate) fn invoke_wasm_and_catch_traps<T>(
 ///   allocated by WebAssembly code and it's relative to the initial stack
 ///   pointer that called into wasm.
 ///
-/// * Stack canaries for externref gc tracing. Currently the implementation
-///   relies on walking frames but the stack walker isn't always 100% reliable,
-///   so a canary is used to ensure that if the canary is seen then it's
-///   guaranteed all wasm frames have been walked.
-///
 /// This function may fail if the the stack limit can't be set because an
 /// interrupt already happened.
 fn enter_wasm<T>(store: &mut StoreContextMut<'_, T>) -> Option<usize> {
-    // If this is a recursive call, e.g. our stack canary is already set, then
-    // we may be able to skip this function.
-    //
-    // For synchronous stores there's nothing else to do because all wasm calls
-    // happen synchronously and on the same stack. This means that the previous
-    // stack limit will suffice for the next recursive call.
+    // TODO FITZGEN: Do we need a non-stack canary way to detect re-entrancy and
+    // early exit here?
     //
-    // For asynchronous stores then each call happens on a separate native
-    // stack. This means that the previous stack limit is no longer relevant
-    // because we're on a separate stack. In this situation we need to
-    // update the stack limit, but we don't need to update the gc stack canary
-    // in this situation.
-    if store
-        .0
-        .externref_activations_table()
-        .stack_canary()
-        .is_some()
-        && !store.0.async_support()
-    {
-        return None;
-    }
+    // // If this is a recursive call, e.g. our stack canary is already set, then
+    // // we may be able to skip this function.
+    // //
+    // // For synchronous stores there's nothing else to do because all wasm calls
+    // // happen synchronously and on the same stack. This means that the previous
+    // // stack limit will suffice for the next recursive call.
+    // //
+    // // For asynchronous stores then each call happens on a separate native
+    // // stack. This means that the previous stack limit is no longer relevant
+    // // because we're on a separate stack. In this situation we need to
+    // // update the stack limit, but we don't need to update the gc stack canary
+    // // in this situation.
+    // if store
+    //     .0
+    //     .externref_activations_table()
+    //     .stack_canary()
+    //     .is_some()
+    //     && !store.0.async_support()
+    // {
+    //     return None;
+    // }
```

Comment:
We'll actually need to replace this check with some other boolean in a `store` since the non-recursive behavior here was necessary to implement the currently desired semantics (stack limit set only once) and we just piggy-backed on the canary being set.

Reply:
Any suggestions of how to implement this check? Do we have anything already that we can piggy back on, the way we did with the stack canary, or do we need a new thing?

Reply:
Sorry I meant to answer this earlier and forgot. I think `stack_limit` could be suitable for this in the `VMRuntimeLimits`. If it's `usize::MAX` then it should be reset, otherwise it should be not tampered with.